# 1. Import Data & Libraries

In [1]:
import pandas as pd
import numpy as np

# from ydata_profiling import ProfileReport

# viz
import missingno as msno
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px


#%matplotlib inline

from sklearn.preprocessing import OneHotEncoder

pd.set_option('display.max_columns', None)
# Suppress Warnings
import warnings
warnings.filterwarnings("ignore")

import functions as f

**Data**

In [2]:
df = pd.read_csv('./Case1_HotelCustomerSegmentation.csv', 
                 sep = ';', index_col = 'ID')
df

Nationality   Age  DaysSinceCreation  \
ID                                            
1              PRT  52.0                440   
2              PRT   NaN               1385   
3              DEU  32.0               1385   
4              FRA  61.0               1385   
5              FRA  52.0               1385   
...            ...   ...                ...   
111729         DEU  31.0                 36   
111730         BRA  43.0                 36   
111731         BRA  37.0                 36   
111732         DEU  48.0                 36   
111733         DEU  48.0                 36   

                                                 NameHash  \
ID                                                          
1       0x2C371FD6CE12936774A139FD7430C624F1C4D5109CE6...   
2       0x198CDB98BF37B6E23F9548C56A88B00912D65A9AA0D6...   
3       0xDA46E62F66936284DF2844EC4FC542D0DAD780C0EE0C...   
4       0xC45D4CD22C58FDC5FD0F95315F6EFA5A6E7149187D49...   
5       0xD2E3D5BFCA141865669F98D64CDA85AD04DEFF47F8A0...   
...                                                   ...   
111729  0x2F59F0E86596B861C3303585FDB7F090B1B2D70442C5...   
111730  0x2708B9F11C95F384129152CDF0830B566F02D42B87AC...   
111731  0xEF5F6AAA6120F2AC49179E16FD12477C243A8F4FC504...   
111732  0xC70AE03F615CB96B0C84CF29419C2DDFE8EC64B77104...   
111733  0xE52EE501843E3108F20BB8F2EC180D8654DF1424EFA2...   

                                                DocIDHash  AverageLeadTime  \
ID                                                                           
1       0x434FD3D59469C73AFEA087017FAF8CA2296493AEABDE...               59   
2       0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...               61   
3       0x27F5DF762CCDA622C752CCDA45794923BED9F1B66300...                0   
4       0x8E59572913BB9B1E6CAA12FA2C8B7BF387B1D1F3432E...               93   
5       0x42BDEE0E05A9441C94147076EDDCC47E604DA5447DD4...                0   
...                                                   ...              ...   
111729  0x86AC02B9393E7F58354BA4193D792C07DB91D4BE27C7...                0   
111730  0xE87DEB08B0D7D0BDC590949FF04AAA893018BD8EB714...              170   
111731  0x74B9B5A554F5F660371E56D4CE4C1D4C3DC80AE71D68...                0   
111732  0xF589C016E6988AECD3E3BE793E7D606BBDB6B6D4795C...               66   
111733  0xD539E89183AB389864B7A24B3C1FBC1FE20A40ED02E7...                0   

        LodgingRevenue  OtherRevenue  BookingsCanceled  BookingsNoShowed  \
ID                                                                         
1               292.00          82.3                 1                 0   
2               280.00          53.0                 0                 0   
3                 0.00           0.0                 0                 0   
4               240.00          60.0                 0                 0   
5                 0.00           0.0                 0                 0   
...                ...           ...               ...               ...   
111729            0.00           0.0                 0                 0   
111730          755.25          20.0                 0                 0   
111731            0.00           0.0                 0                 0   
111732          708.00         185.0                 0                 0   
111733            0.00           0.0                 0                 0   

        BookingsCheckedIn  PersonsNights  RoomNights    DistributionChannel  \
ID                                                                            
1                       2              6           4              Corporate   
2                       1             10           5  Travel Agent/Operator   
3                       0              0           0  Travel Agent/Operator   
4                       1             10           5  Travel Agent/Operator   
5                       0              0           0  Travel Agent/Operator   
...                   ...            ...    

# 3. Data Cleaning and Preprocessing

## 3.1 Duplicates

As previously identified, 111 duplicate value were found therefore we proceed to removing them from the dataset.

In [3]:
df.drop_duplicates(inplace = True)

<br>

## 3.2 Fixing inconsistencies

**Customers' age**

To ensure data consistency, we will treat ages below 16 and 90 or above as inconsistent and mark them as missing. This will help improve the quality of the booking data.

In [4]:
df.loc[(df['Age'] < 16) | (df['Age'] > 90), 'Age'] = np.nan

<br>

**Hash of customer identification number**

For this feature, we considered merging records based on `NameHash`, `DocIDHash` and `DistributionChannel`, this will allow to keep a record for the same customer. This approach ensures that each customer is represented by a single record while still accounting for repeated DocIDHash values. These duplicates may arise due to bookings made through different channels or corporate reservations where multiple names are linked to the same document (DEPENDENDO DO QUE O PROF DIZER TALVEZ AQUI POSSAMOS TB DIZER QUE ISTO PODERÁ ACONTECER PQ SÃO OS FILHOS, POR ISSO APARECE NOMES DIFERENTES, MAS COM O MESMO ID).

In [5]:
df2 = df.dropna(subset=['DocIDHash'])
df2

Nationality   Age  DaysSinceCreation  \
ID                                            
1              PRT  52.0                440   
2              PRT   NaN               1385   
3              DEU  32.0               1385   
4              FRA  61.0               1385   
5              FRA  52.0               1385   
...            ...   ...                ...   
111729         DEU  31.0                 36   
111730         BRA  43.0                 36   
111731         BRA  37.0                 36   
111732         DEU  48.0                 36   
111733         DEU  48.0                 36   

                                                 NameHash  \
ID                                                          
1       0x2C371FD6CE12936774A139FD7430C624F1C4D5109CE6...   
2       0x198CDB98BF37B6E23F9548C56A88B00912D65A9AA0D6...   
3       0xDA46E62F66936284DF2844EC4FC542D0DAD780C0EE0C...   
4       0xC45D4CD22C58FDC5FD0F95315F6EFA5A6E7149187D49...   
5       0xD2E3D5BFCA141865669F98D64CDA85AD04DEFF47F8A0...   
...                                                   ...   
111729  0x2F59F0E86596B861C3303585FDB7F090B1B2D70442C5...   
111730  0x2708B9F11C95F384129152CDF0830B566F02D42B87AC...   
111731  0xEF5F6AAA6120F2AC49179E16FD12477C243A8F4FC504...   
111732  0xC70AE03F615CB96B0C84CF29419C2DDFE8EC64B77104...   
111733  0xE52EE501843E3108F20BB8F2EC180D8654DF1424EFA2...   

                                                DocIDHash  AverageLeadTime  \
ID                                                                           
1       0x434FD3D59469C73AFEA087017FAF8CA2296493AEABDE...               59   
2       0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...               61   
3       0x27F5DF762CCDA622C752CCDA45794923BED9F1B66300...                0   
4       0x8E59572913BB9B1E6CAA12FA2C8B7BF387B1D1F3432E...               93   
5       0x42BDEE0E05A9441C94147076EDDCC47E604DA5447DD4...                0   
...                                                   ...              ...   
111729  0x86AC02B9393E7F58354BA4193D792C07DB91D4BE27C7...                0   
111730  0xE87DEB08B0D7D0BDC590949FF04AAA893018BD8EB714...              170   
111731  0x74B9B5A554F5F660371E56D4CE4C1D4C3DC80AE71D68...                0   
111732  0xF589C016E6988AECD3E3BE793E7D606BBDB6B6D4795C...               66   
111733  0xD539E89183AB389864B7A24B3C1FBC1FE20A40ED02E7...                0   

        LodgingRevenue  OtherRevenue  BookingsCanceled  BookingsNoShowed  \
ID                                                                         
1               292.00          82.3                 1                 0   
2               280.00          53.0                 0                 0   
3                 0.00           0.0                 0                 0   
4               240.00          60.0                 0                 0   
5                 0.00           0.0                 0                 0   
...                ...           ...               ...               ...   
111729            0.00           0.0                 0                 0   
111730          755.25          20.0                 0                 0   
111731            0.00           0.0                 0                 0   
111732          708.00         185.0                 0                 0   
111733            0.00           0.0                 0                 0   

        BookingsCheckedIn  PersonsNights  RoomNights    DistributionChannel  \
ID                                                                            
1                       2              6           4              Corporate   
2                       1             10           5  Travel Agent/Operator   
3                       0              0           0  Travel Agent/Operator   
4                       1             10           5  Travel Agent/Operator   
5                       0              0           0  Travel Agent/Operator   
...                   ...            ...    

In [6]:
bookings = f.aggregation(df2)

In [7]:
bookings

DocIDHash  \
0       0x00006E606D590F336C289D2CC995A0C726878BCAC001...   
1       0x00006E606D590F336C289D2CC995A0C726878BCAC001...   
2       0x0000E25BC06E730F6DDD6BD53CBFE05BABD2B0182019...   
3       0x0001425DE03D204F30757BA2E50DBD0486ED254E1E6F...   
4       0x0002AE77553DBEBD734F0C644E6BE142EA0B4C5D5D83...   
...                                                   ...   
108202  0xFFFB1DBFAB9D8FFDC8602C6D365578FBFF637070F3EB...   
108203  0xFFFB9FE60E0D35B8426F2FB90646D0BA799B803A79C1...   
108204  0xFFFE14C4D0D6520C97125A115CDC705B8A2F0E5133D5...   
108205  0xFFFED0F2BC7D4A410C579CEE04804A70F09C1B21B340...   
108206  0xFFFF697C455DCDFA906747736B03F3A61D20A92C8F74...   

                                                 NameHash  \
0       0x0D54EA6B14A0574CAA2D3802565441048C5C831EB035...   
1       0x0D54EA6B14A0574CAA2D3802565441048C5C831EB035...   
2       0xF33C33AE994E1AE82D574B2BF23945DC93A9D9CC89FE...   
3       0x42FC523622627DB3F09A6A544335CDAF2EDA9C57BA61...   
4       0xF47414CC56472838252EE0BB9C0639AEA24312985ECA...   
...                                                   ...   
108202  0xC96442DFC501BAF021CB0FE9CC15C1C22A98B8672FAF...   
108203  0x342C21783C3293511503091DCBCD4EB112AB13C2CD46...   
108204  0xFD4A86C21CE0E27CFDF3ED6BFAF83F6B288657F972A2...   
108205  0x5CF6B8387455E82BC6A2BCD111C0117527AEA2B37BDC...   
108206  0xFADD4AC7A38534209C862EF3C9FC47C74509433C8A23...   

          DistributionChannel Nationality   Age  DaysSinceCreation  \
0                      Direct         FRA  60.0                479   
1       Travel Agent/Operator         FRA  60.0                480   
2       Travel Agent/Operator         DEU  54.0               1197   
3                      Direct         USA  34.0                120   
4       Travel Agent/Operator         IND   NaN                569   
...                       ...         ...   ...                ...   
108202                 Direct         USA  49.0                123   
108203                 Direct         CAN  41.0                783   
108204  Travel Agent/Operator         FRA  19.0                 46   
108205  Travel Agent/Operator         IRL  26.0                507   
108206  Travel Agent/Operator         ITA  60.0                 55   

        AverageLeadTime  LodgingRevenue  OtherRevenue  BookingsCanceled  \
0                   1.0           177.0          14.0                 0   
1                 127.0           318.0         148.0                 0   
2                  14.0           695.0         100.0                 0   
3                   0.0             0.0           0.0                 0   
4                   0.0             0.0           0.0                 0   
...                 ...             ...           ...               ...   
108202              0.0             0.0           0.0                 0   
108203             58.0           218.0          14.0                 0   
108204             89.0           911.7          31.0                 0   
108205              0.0             0.0           0.0                 0   
108206            176.0           730.8          96.0                 0   

        BookingsNoShowed  BookingsCheckedIn  PersonsNights  RoomNights  \
0                      0                  1              2           1   
1                      0                  1              4           2   
2                      0                  1             10           5   
3                      0                  0              0           0   
4                      0                  0              0           0   
...                  ...                ...            ...         ...   
108202                 0                  0              0           0   
108203                 0                  1              2           2   
108204                 0                  1             14           7   
108205                 0                  0              0           0   
108206                 0  

In [8]:
bookings['DocIDHash'].value_counts()

DocIDHash
0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855    2834
0x1B16B1DF538BA12DC3F97EDBB85CAA7050D46C148134290FEBA80F8236C83DB9      10
0xA486FBACF4B4E5537B026743E3FDFE571D716839E758236F42950A61FE6B922B       9
0x3856085146F7BC27BD07BFC4CA1991ED4E65E179D7BDB7DBBA7E32620809C799       7
0xE76926D097F612D921A7494C6B18247BF7DBA28D0679960312DF2AD5DB7F37BD       7
                                                                      ... 
0x55A9187065EFE381F6211B1B98952620B8A784A29DE449FEDA41712064A4AA42       1
0x55A79C0BD3226CA1E1723B0C794192042D85B772D2A667DE2C59F9D25A0B648F       1
0x55A769E36CA3B9C45FF16B6A761BF1CBE8A486882B3AB208395A4686548EC8E4       1
0x55A720611D39B0072081BF4F259DA05CCCF26D4D3C12A17F9EA1DB6E241415A7       1
0xFFFF697C455DCDFA906747736B03F3A61D20A92C8F741ED432C03B2E7E5DD6AD       1
Name: count, Length: 103480, dtype: int64

From the output above, it is clear that we were able to reduce the number of records that were displaying duplicated information about the same customer.

<br>

**Number of days in advance that the customer makes the booking before the arrival date**

We identified a negative value in the `AverageLeadTime` feature, which we consider to be an error. As a result, we will treat these values as missing and proceed with imputation. Given that most customers are not from Portugal, imputing values like 0 or 1 wouldn't be realistic, as it is unlikely that customers would arrive without sufficient travel time. Instead, we will use K-nearest neighbors for imputation, which provides a more reasonable approach for estimating these missing values.

In [15]:
bookings[bookings['AverageLeadTime'] == -1].head()

DocIDHash  \
17604  0x2A1EA7292AF2D9BB083726E0961A38A54950480886BA...   
27883  0x43126A8B69B8B1E3C1D4513EE586ABE2910947964929...   
31948  0x4CDEF046A3A97AAC6BD603F2250C8ACCB068C39995C4...   
35137  0x54D0B32490EA6DAB16F26AFBDAD5037A89E1BDA62B54...   
53275  0x814B89362FCA5770AEDC4485136D7F673F78793CE218...   

                                                NameHash  \
17604  0x4567E92118939CF20A432369E32BA8959502682AD473...   
27883  0xCD4216685825A5A65B100A4F8F19366A23B2BDDE8E29...   
31948  0x61B41490861314B7148F76EE7D7E17A240A86B45259D...   
35137  0xBBBA5623819CD060BB49D9D700348FFEF885F825D019...   
53275  0x7DD53BA3E94C8D8B171B9D50041DC3A23E6F4752A3A8...   

         DistributionChannel Nationality   Age  DaysSinceCreation  \
17604  Travel Agent/Operator         USA  70.0                468   
27883  Travel Agent/Operator         BRA  31.0                394   
31948  Travel Agent/Operator         CHN  33.0                558   
35137  Travel Agent/Operator         ESP  46.0                485   
53275  Travel Agent/Operator         FRA  61.0                499   

       AverageLeadTime  LodgingRevenue  OtherRevenue  BookingsCanceled  \
17604             -1.0            95.2           2.0                 0   
27883             -1.0           209.0           2.0                 0   
31948             -1.0           129.0           2.0                 0   
35137             -1.0           159.0           2.0                 0   
53275             -1.0           149.0           2.0                 0   

       BookingsNoShowed  BookingsCheckedIn  PersonsNights  RoomNights  \
17604                 0                  1              2           1   
27883                 0                  1              2           1   
31948                 0                  1              2           1   
35137                 0                  1              2           1   
53275                 0                  1              2           1   

      MarketSegment  SRHighFloor  SRLowFloor  SRAccessibleRoom  SRMediumFloor  \
17604         Other            0           0                 0              0   
27883         Other            0           0                 0              0   
31948         Other            0           0                 0              0   
35137         Other            0           0                 0              0   
53275         Other            0           0                 0              0   

       SRBathtub  SRShower  SRCrib  SRKingSizeBed  SRTwinBed  SRNearElevator  \
17604          0         0       0              0          0               0   
27883          0         0       0              0          0               0   
31948          0         0       0              0          0               0   
35137          0         0       0              0          0               0   
53275          0         0       0              0          0               0   

       SRAwayFromElevator  SRNoAlcoholInMiniBar  SRQuietRoom  
17604                   0                     0            0  
27883                   0                     0            0  
31948                   0                     0            0  
35137                   0                     0            0  
53275                   0                     0            0

In [16]:
bookings.loc[bookings['AverageLeadTime'] == -1, 'AverageLeadTime'] = np.nan

<br>

**Total amount of lodging revenue paid by the customer so far**

From previous insights, we identified customers who have not spend in any hotel amenities including both room charges and other expenses such as the bar. Since these customers do not contribute to the company's revenue, we have decided to exclude them from the dataset.

In [27]:
print(f'{bookings[(bookings['LodgingRevenue'] == 0) & (bookings['OtherRevenue'] == 0)].shape[0]} customers did not generate revenue to the hotel.')

31898 customers did not generate revenue to the hotel.


In [28]:
bookings.drop(bookings[(bookings['LodgingRevenue'] == 0) & (bookings['OtherRevenue'] == 0)].index, inplace=True)

31898 records were removed from the dataset.

<br>

**Booking check-ins, revenues and persons/rooms/nights**

We have identified some incoherences, namely:

- there we customers with `BookingsCheckedIn` value equal to 0, meaning that these are not customers;
- there was one customer customer who made purchases for both rooms and other amenities but reported no cancellations, no-show, or check-in activity. This is inconsistent, as it is unlikely that a customer who made such charges did not at least check in to the hotel. Therefore, this data anomaly suggests an error in the record.

In [30]:
bookings[bookings['BookingsCheckedIn']==0]

Empty DataFrame
Columns: [DocIDHash, NameHash, DistributionChannel, Nationality, Age, DaysSinceCreation, AverageLeadTime, LodgingRevenue, OtherRevenue, BookingsCanceled, BookingsNoShowed, BookingsCheckedIn, PersonsNights, RoomNights, MarketSegment, SRHighFloor, SRLowFloor, SRAccessibleRoom, SRMediumFloor, SRBathtub, SRShower, SRCrib, SRKingSizeBed, SRTwinBed, SRNearElevator, SRAwayFromElevator, SRNoAlcoholInMiniBar, SRQuietRoom]
Index: []

Based on the output, we can confirm that this issue no longer occurs, as we had removed records from the dataset.

### isto acontece, o DocIDHash do individuo está missing

In [33]:
bookings[(bookings['BookingsCheckedIn'] == 0) & ((bookings['LodgingRevenue'] != 0) | (bookings['OtherRevenue'] != 0))]

Empty DataFrame
Columns: [DocIDHash, NameHash, DistributionChannel, Nationality, Age, DaysSinceCreation, AverageLeadTime, LodgingRevenue, OtherRevenue, BookingsCanceled, BookingsNoShowed, BookingsCheckedIn, PersonsNights, RoomNights, MarketSegment, SRHighFloor, SRLowFloor, SRAccessibleRoom, SRMediumFloor, SRBathtub, SRShower, SRCrib, SRKingSizeBed, SRTwinBed, SRNearElevator, SRAwayFromElevator, SRNoAlcoholInMiniBar, SRQuietRoom]
Index: []

# Aqui não sei muito bem o que fazer

<br>

**Number of check-in higher than the number of persons/nights at the hotel and rooms**

In [34]:
bookings[bookings['BookingsCheckedIn']>bookings['PersonsNights']]

DocIDHash  \
2080   0x050B3FEBB73D2924810D9C068846F14550B03BCFDC7C...   
38782  0x5DC0150F4FDDED43B313C3782560597AC6665A129EF5...   
43009  0x67D8A25C96313AEF8A5842A87F2F6AB29DE01DAC80BE...   
53376  0x8186236B5A280E296DC622D0B419BBEBB5BC8E1D4338...   
93943  0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
96155  0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
97109  0xE5123E18BD091C1108E145BE447859B469199AAAC5D7...   

                                                NameHash  \
2080   0xFC913F891030DD576FDA085E7859D44AB7EE8B000D59...   
38782  0x7A8179ADB882FC633830D7EABDCD64D41B3CD539FEA2...   
43009  0x65E5986F79B75C8A171FEEA2E534AAD806D5ACB083D6...   
53376  0x0596170AE19299CB3CEBA98BC2FAF9BF5333031C6A32...   
93943  0x14A77C11BA7C3B13207E72D4BA395580338D206DCA65...   
96155  0xDCA87EABFF56A1A313E731ED676F25361722AB5AE790...   
97109  0x9EEAA6DAE9241E36FF1A13DF7ED180939C37FC54E6BA...   

         DistributionChannel Nationality   Age  DaysSinceCreation  \
2080   Travel Agent/Operator         DEU  52.0                214   
38782              Corporate         PRT  46.0                466   
43009  Travel Agent/Operator         NLD  33.0                641   
53376  Travel Agent/Operator         BRA  35.0                336   
93943  Travel Agent/Operator         ESP   NaN                377   
96155  Travel Agent/Operator         PRT   NaN               1314   
97109  Travel Agent/Operator         FIN  51.0                729   

       AverageLeadTime  LodgingRevenue  OtherRevenue  BookingsCanceled  \
2080              36.0          2514.0         103.5                 0   
38782              3.0           269.0           7.0                 0   
43009             41.0           267.0         109.0                 0   
53376             63.0           242.0          24.0                 0   
93943             69.0           488.0          49.0                 0   
96155              8.0             0.0        5105.5                 0   
97109              9.0           798.0         177.5                 0   

       BookingsNoShowed  BookingsCheckedIn  PersonsNights  RoomNights  \
2080                  0                  1              0           2   
38782                 1                  2              1           2   
43009                 0                  1              0           4   
53376                 0                  1              0           2   
93943                 0                  2              1           4   
96155                 0                  3              2          12   
97109                 0                  2              1           1   

               MarketSegment  SRHighFloor  SRLowFloor  SRAccessibleRoom  \
2080   Travel Agent/Operator            0           0                 0   
38782               Aviation            0           0                 0   
43009                 Groups            0           0                 0   
53376                  Other            0           0                 0   
93943                 Groups            0           0                 0   
96155              Corporate            0           0                 0   
97109                  Other            1           0                 0   

       SRMediumFloor  SRBathtub  SRShower  SRCrib  SRKingSizeBed  SRTwinBed  \
2080               0          0         0       0              0          0   
38782              0          0         0       0              1          0   
43009              0          0         0       0              0          1   
53376              0          0         0       1              1          0   
93943              0          0         0       0              0          0   
96155              0          0         0       0              0          0   
97109              0          0         0       0              1          0   

       SRNearElevator  SRAwayFromElevator  SRNoAlcoholInMiniBar  SRQuietRoom  
2080                0                   0

In [35]:
bookings[bookings['PersonsNights']<bookings['RoomNights']]

DocIDHash  \
2080    0x050B3FEBB73D2924810D9C068846F14550B03BCFDC7C...   
3319    0x08007328E88BA93B551B9F9F056AAFE2C00E94CCEE69...   
16583   0x27A6504F8E07DC750D565994E86CB3837146A501E609...   
17079   0x28D83B97B0D5B78BCBA1C36FDF79FF276A1E32C9CB5C...   
18019   0x2B17E9D2CCEF2EA0FE752EE345BEDFB06741FFC8ECEC...   
26462   0x3FAE56E6A0168E9130F12E58CF45CDAAE18477EF809C...   
27462   0x420A533D66150B9173FAB8F5AF0DB8C54EBDFA001663...   
32252   0x4DA48EED943299176444C244EAA74FACB04E8E66DE82...   
38782   0x5DC0150F4FDDED43B313C3782560597AC6665A129EF5...   
41903   0x6541AF2C1E55C2D3EC17E89DEDD22FE0EBB47AB92305...   
43009   0x67D8A25C96313AEF8A5842A87F2F6AB29DE01DAC80BE...   
53376   0x8186236B5A280E296DC622D0B419BBEBB5BC8E1D4338...   
54602   0x849546932E0ABE7515ECA03635FD1334BBBCC41E0AFD...   
64855   0x9E0C31D4C3A418F0B4E063504160A0B33D51C6ABA0A4...   
66034   0xA0EA86D350113A03143831212044DE6CE61C2484CB18...   
67701   0xA508EDBFE269E231546629F11587FA2AAD4DE4851A31...   
78386   0xBEDD3BDEFDC2F17B6B52D1133594F2A0A60D7D665DEB...   
82261   0xC80D1E0E22E0899087DECE589FC5BAE8FB1CAC4CED2D...   
86969   0xD37A405EAA2BF8D23CB44A9AE3C2A7BEC4C5A726BEE2...   
93240   0xE2836653FBF39129A11F69F1FC6635CCCD8FA5C7A5E1...   
93943   0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
95542   0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
96122   0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
96155   0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
104586  0xF728641F469966A5AA8906B76CC8C3083A0BDEC9F24F...   
107824  0xFF1985F366C218F0D7D8280480B6B8AD82CAF6092799...   

                                                 NameHash  \
2080    0xFC913F891030DD576FDA085E7859D44AB7EE8B000D59...   
3319    0x84DF93E126886E0003E45A8B0A1127EDDDA6A3BEC276...   
16583   0x53FB0EB1CE1FD50424F79500C54F8149081D5A3D82F6...   
17079   0xFB5F026540ECB47362289BB2507D92BED41D80F632F6...   
18019   0xFB64B4B6AB53A6A549A620009CA24A1C3A668A460D87...   
26462   0x04F37DBBE0138585CD31BC820E2C780F992915176507...   
27462   0xD0C14D44BB72D7FA9673B2EE2EE9C8021EBDF81A7EF0...   
32252   0xE779FF69A9BC8AA6FE091759A095386C0E2A8F680DA8...   
38782   0x7A8179ADB882FC633830D7EABDCD64D41B3CD539FEA2...   
41903   0xF4F7ED49904EB1F7A72F33A071F96CE533019C81DF96...   
43009   0x65E5986F79B75C8A171FEEA2E534AAD806D5ACB083D6...   
53376   0x0596170AE19299CB3CEBA98BC2FAF9BF5333031C6A32...   
54602   0x3CBB385BBFE0BCBAC87A0F4FE0BFA648DE33905384F4...   
64855   0x27749A514C58DA6E2F5544303DD02E378631187B08EF...   
66034   0x82FDEEE5403B80470AABEC952A25DB7D37BEF253BD5C...   
67701   0xDF5D8EAC38850FC98D7D5822110B10A44662D9D92279...   
78386   0x5015AE870F1DEECA791D3744CD4ED3CAA95B22C43A5B...   
82261   0x2D9E296C60A6A2977D4C4812CB560A443966E5B9FBAE...   
86969   0x257B8DC0708BF75D2A3869648C15B36317B272A1D69B...   
93240   0x4FC57FA3A9C4A4096E55BFD5AACE40D3627748CB65E4...   
93943   0x14A77C11BA7C3B13207E72D4BA395580338D206DCA65...   
95542   0xA54BF56198FA66C8414FE1AEB3680DC262B8EE9ACC21...   
96122   0xD9A501DB0F0303E3FD9B2FCD27B28411ACE0486F77C0...   
96155   0xDCA87EABFF56A1A313E731ED676F25361722AB5AE790...   
104586  0x382BF2A6BF57870D9831631A7ACA7B30CB42FE9FEEDF...   
107824  0xF2A1083BA0F2D0B18B86816CD820DAD61C3818A4A828...   

          DistributionChannel Nationality   Age  DaysSinceCreation  \
2080    Travel Agent/Operator         DEU  52.0                214   
3319    Travel Agent/Operator         PRT  45.0               1288   
16583   Travel Agent/Operator         ESP  50.0                736   
17079                  Direct         PRT  34.0                489   
18019               Corporate         PRT  48.0               1323   
26462               Corporate         GRC  52.0                901   
27462                  Direct         USA  67.0                261   
32252               Corporate         PRT  66.0               1189   
38782               Corporate         PRT  46.0                466   
41903                  Direct         PRT  49.0     

<br>

**Customers under age when they registered**

It is also important to check if exists customers who have created an account when they were less than 16.

In [38]:
bookings[bookings['Age'] - (bookings['DaysSinceCreation'] / 365) < 16]

DocIDHash  \
228     0x00894B7E1CABFA8EED3F3EBD07775EC805097E4C8731...   
593     0x017761CDC0557FA4B549E9B6E57EFF9F2D9999CDD0DF...   
625     0x018BC244424177263A49E155E0155703A55EB5DE5688...   
986     0x025918E07BFFA32395D931CAD20281E0F2B7306F37BF...   
1045    0x0283C4490521CA28CAB3399496A12119EBF56034E47D...   
...                                                   ...   
107720  0xFED7CA45EDCD56E0569A1DD3299896E10FF6BA2D0B53...   
107778  0xFEFB52E2FC271B128FD396F95C9798E06DCDE9B8F317...   
107802  0xFF091167E8C620839C5886914BA70DFDBA507096579E...   
107911  0xFF51F390B351649222FFE9FEF664AC9182C834A42E0C...   
108118  0xFFCB925A6F18EE11BA4F045ABA25A15C9570F433B2FB...   

                                                 NameHash  \
228     0x5EB470337A85ABCB5B93641CCF89831F0132CE4272B0...   
593     0x0BD21A30BDA465A795AA961D3DFDB180904CB786E567...   
625     0x463B57C1BF57C60D45AE958F37608B1CE2EDFF695147...   
986     0x44538360D90434312302A54D95005529CDA5713B75CB...   
1045    0xC8BCD3E725F9AC4133E2560781DEB0B448A0F21193CC...   
...                                                   ...   
107720  0x0A48007FB86C37A62DC1A25E16CACA32CD73E7593921...   
107778  0xE7E2BD819585F3A1DAAE61012B860291848AEA985E70...   
107802  0x28D865CAA16F0FA0992FD6AAA7391D90A7DD032ABBB7...   
107911  0x7B44D5B34FC5911E1303D123C73059EB1B0F89096AD6...   
108118  0x132922896CCD514B35138A018C937C03948CF4587A98...   

          DistributionChannel Nationality   Age  DaysSinceCreation  \
228                    Direct         FRA  17.0                977   
593     Travel Agent/Operator         FRA  16.0                909   
625     Travel Agent/Operator         NLD  16.0                366   
986     Travel Agent/Operator         ESP  16.0                676   
1045                   Direct         DEU  19.0               1301   
...                       ...         ...   ...                ...   
107720  Travel Agent/Operator         NLD  19.0               1163   
107778  Travel Agent/Operator         UMI  16.0                462   
107802  Travel Agent/Operator         GBR  17.0               1167   
107911  Travel Agent/Operator         BEL  17.0                796   
108118  Travel Agent/Operator         CHE  17.0                961   

        AverageLeadTime  LodgingRevenue  OtherRevenue  BookingsCanceled  \
228                 9.0         1021.20        239.94                 0   
593               114.0          456.00         79.50                 0   
625                12.0          497.00          6.00                 0   
986                 9.0          342.00         40.00                 0   
1045              130.0          283.50         56.00                 0   
...                 ...             ...           ...               ...   
107720            183.0          239.40         42.00                 0   
107778             10.0         1199.66         69.50                 0   
107802            164.0          189.00         62.00                 0   
107911            203.0          337.50         45.50                 0   
108118            134.0          687.96         66.00                 0   

        BookingsNoShowed  BookingsCheckedIn  PersonsNights  RoomNights  \
228                    0                  1              8           4   
593                    0                  1              8           4   
625                    0                  1              6           3   
986                    0                  1              6           2   
1045                   0                  1              8           4   
...                  ...                ...            ...         ...   
107720                 0                  1              6           2   
107778                 0                  1              7           7   
107802                 0                  1              4           2   
107911                 0                  1             10           5   
108118                 0  